internal/controller/controllers/crd_utils.go

```diff
@@ -0,0 +1,69 @@
+package controllers
+
+import (
+	"context"
+
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const APIVersion = "adi.io.my.domain/v1alpha1"
+
+type CRDUtils struct {
+	client client.Client
+}
+
+func NewCRDUtils(client client.Client) *CRDUtils {
+	return &CRDUtils{client: client}
+}
+
+func (u *CRDUtils) CreateAgentCR(ctx context.Context, hostId, clusterNamespace, clusterName string) error {
+
+	if clusterNamespace == "" {
+		return nil
+	}
+
+	host := &adiiov1alpha1.Agent{}
+	namespacedName := types.NamespacedName{
+		Namespace: clusterNamespace,
+		Name:      hostId,
+	}
+	err := u.client.Get(ctx, namespacedName, host)
+	if err == nil {
+		return nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		host := &adiiov1alpha1.Agent{
+			Spec: adiiov1alpha1.AgentSpec{
+				ClusterDeploymentName: &adiiov1alpha1.ClusterReference{
+					Name:      clusterName,
+					Namespace: clusterNamespace,
+				},
+			},
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Agent",
+				APIVersion: APIVersion,
+			},
```

Comment:
This should not be necessary. The Scheme that the client uses keeps track of a mapping between GVK and a go struct. As long as the Agent type is registered with the Scheme, this shouldn't be needed.

Reply:
Done